### PR TITLE
llext: add STT_OBJECT relocation

### LINKED ISSUE #67701: llext: STT_OBJECT relocations incorrectly ignored
STT_OBJECT symbol relocations are ignored silently currently and this leads to confusing bugs where a global (non-static) is declared in an extension leading to invalid memory addresses to the global symbol once the extension loads on Arm.

Primarily this is because on Arm we are doing static linking currently I believe.

So don't ignore STT_OBJECT relocations, and certainly don't do it silently

Thanks @laurenmurphyx64 for hunting this down

### PR DESCRIPTION
**Pull request by laurenmurphyx64**
Adds STT_OBJECT relocation.

Fixes #67701.

You can test with the attached patch (it's a patch, just renamed) [0001-STT_OBJECT_test.patch.log](https://github.com/zephyrproject-rtos/zephyr/files/14397787/0001-STT_OBJECT_test.patch.log) and the `tests/subsys/llext/hello_world` test.

### CODE REVIEW

File: subsys/llext/llext.c [hunk 1]
```diff
@@ -776,6 +777,7 @@ static int llext_link(struct llext_loader *ldr, struct llext *ext, bool do_local
 				LOG_INF("found section symbol %s addr 0x%lx", name, link_addr);
 			} else {
 				/* Nothing to relocate here */
+				LOG_INF("not relocated");
```
Comment (teburd) [inline subsys/llext/llext.c:780]: I'd say this should likely be a LOG_DBG

**Review by teburd — COMMENTED** [review]
Change looks good, I think it could us with a test case

copying the hello_world extension and updating it to work a little different and do what your patch does with a global (non-static global) variable should do the trick.

The base ztest suite should be the same, and run through all the available test extensions like the simplest hello world printk, this sem give/take with a global, and so on.

File: subsys/llext/llext.c (hunk 1 above)
Comment (laurenmurphyx64) [reply subsys/llext/llext.c]: > The base ztest suite should be the same, and run through all the available test extensions like the simplest hello world printk, this sem give/take with a global, and so on.

@teburd I've put one up, have a look and see if you like it / let me know what other test cases you'd like to see.

**Comment by laurenmurphyx64** [comment]
> The base ztest suite should be the same, and run through all the available test extensions like the simplest hello world printk, this sem give/take with a global, and so on.

@teburd I've put one up, have a look and see if you like it / let me know what other test cases you'd like to see.

**Review by pillo79 — COMMENTED** [review]
Change looks obviously good and thanks for the test code!
 
Personally, I'd like `hello world` to be the simplest possible example. I would therefore create a separate `thread_test` project integrating `CONFIG_USERSPACE` and your new thread test code, and removing the duplicate code from the original `hello_world`. @teburd what's your take?

**Comment by teburd** [comment]
> Change looks obviously good and thanks for the test code!
> 
> Personally, I'd like `hello world` to be the simplest possible example. I would therefore create a separate `thread_test` project integrating `CONFIG_USERSPACE` and your new thread test code, and removing the duplicate code from the original `hello_world`. @teburd what's your take?

This can be done with testcase.yml, consolidating the tests into one test project is helpful honestly. The more of these tests we have that are seperate they tend to add to actual test time primarily from building and flashing rather than actually running test code.


> Change looks obviously good and thanks for the test code!
> 
> Personally, I'd like `hello world` to be the simplest possible example. I would therefore create a separate `thread_test` project integrating `CONFIG_USERSPACE` and your new thread test code, and removing the duplicate code from the original `hello_world`. @teburd what's your take?

The test was named simple to exercise simple extensions and their code paths, this is still a pretty simple extension and I think a worthwhile addition here.

The trouble with making too many separate test applications is the build and flash time grows faster than the time it takes to run the tests by a pretty large margin sometimes, especially for something like this.

If separately compiled tests do become necessary we can use testcase.yaml to decide which extension to test while utilizing all the test code structure still. Personally I don't think it'll be necessary though, the extensions themselves are small.

**Comment by pillo79** [comment]
Thanks for the explanations. 

---

Looked at CI failures - they are due to the `-mcpu=cortex33+nodsp` flag that llext forces when compiling ARM extensions. That has been done in the early commits to force the llext compiler to use a fixed CPU instruction set (I _think_ to disable complex jump opcodes from being emitted?), but it's obviously a hack that should be addressed.

**Comment by cfriedt** [comment]
@laurenmurphyx64 - looks good.

Just one question - do we already have exported syscalls pr exported symbols for `k_sem_init()`, `k_thread_join()`, etc?

There are likely quite a number of thread-related exports that might be necessary in the long run, but if we're exporting `k_thread_create()`, we should probably simultaneously export `k_thread_join()` at least.

Are there any other thread APIs that would be safe to complement `k_thread_create()`? E.g. `k_thread_abort()`?

**Comment by laurenmurphyx64** [comment]
> There are likely quite a number of thread-related exports that might be necessary in the long run, but if we're exporting `k_thread_create()`, we should probably simultaneously export `k_thread_join()` at least.

@cfriedt I'm working on adding k_thread_join, but trying to use it causes faults ("unaligned memory access"). Hopefully k_thread_abort won't do the same thing when I try it

**Comment by teburd** [comment]
Please rebase, should fix CI issues

**Comment by laurenmurphyx64** [comment]
Revised to simplify example to just object relocation to keep it in scope of the PR, threads to come in another PR for a bug report. Should be ready to merge assuming CI succeeds...

File: tests/subsys/llext/simple/src/object_ext.c [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This code demonstrates the use of threads and requires object
```
Comment (teburd) [inline tests/subsys/llext/simple/src/object_ext.c:8]: This comment seems out of date now

File: tests/subsys/llext/simple/src/object_ext.c [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2023 Intel Corporation.
```
Comment (teburd) [inline tests/subsys/llext/simple/src/object_ext.c:2]: 2024 has arrived, I just caught myself today doing this :calendar:

**Review by teburd — DISMISSED** [review]
Minor nits

File: tests/subsys/llext/simple/src/object_ext.c (hunk 2 above)
Comment (laurenmurphyx64) [reply tests/subsys/llext/simple/src/object_ext.c]: Good catch, changed

File: tests/subsys/llext/simple/src/object_ext.c (hunk 3 above)
Comment (laurenmurphyx64) [reply tests/subsys/llext/simple/src/object_ext.c]: Updated :clock3:

**Review by teburd — APPROVED** [review]

**Review by pillo79 — APPROVED** [review]

**Comment by marc-hb** [comment]
This commit https://github.com/zephyrproject-rtos/zephyr/commit/ce243944375e46e94346bfb3dcbe8c499fc4996b ("llext: add object test case") and commit https://github.com/zephyrproject-rtos/zephyr/commit/1408d1e5b8cefcb0f4d52c1ffb41849f70826f39 ("tests: llext: compile architectures not supported yet") conflict with each other. They were submitted concurrently, tested separately and merged around the same time. All PRs fail now :-(

Please review hotfix #70406 urgently.

**Pull request merged**